File: benchmark_release/public/case_000017/src/app/services/audit_service.py
"""
Audit logging service — records significant user and admin actions.
"""
from flask import g, request


def get_db():
    return g.db


def log_action(user_id, action, target_type=None, target_id=None, detail=None):
    """Append a row to the audit_log table."""
    db = get_db()
    ip = request.remote_addr if request else None
    try:
        db.execute(
            'INSERT INTO audit_log (user_id, action, target_type, target_id, detail, ip_addr) '
            'VALUES (?,?,?,?,?,?)',
            (user_id, action, target_type, target_id, detail, ip)
        )
        db.commit()
    except Exception:
        pass  # audit failure must never break the main request path


def recent_actions(limit=50):
    db = get_db()
    rows = db.execute(
        'SELECT a.*, u.username FROM audit_log a '
        'LEFT JOIN users u ON a.user_id = u.id '
        'ORDER BY a.created_at DESC LIMIT ?',
        (limit,)
    ).fetchall()
    return [dict(r) for r in rows]

File: benchmark_controls_release/public/case_000020_clean/src/app/services/auditService.js
'use strict';

const { db } = require('../db');

function log(userId, action, detail = '') {
  db.prepare(
    'INSERT INTO audit_log (user_id, action, detail) VALUES (?, ?, ?)'
  ).run(userId, action, detail);
}

function getForUser(userId, limit = 20) {
  return db.prepare(
    'SELECT * FROM audit_log WHERE user_id = ? ORDER BY created_at DESC LIMIT ?'
  ).all(userId, limit);
}

function getAll({ page = 1, pageSize = 25 } = {}) {
  const offset = (page - 1) * pageSize;
  const rows = db.prepare(
    'SELECT al.*, u.username FROM audit_log al LEFT JOIN users u ON u.id = al.user_id ORDER BY al.created_at DESC LIMIT ? OFFSET ?'
  ).all(pageSize, offset);
  const total = db.prepare('SELECT COUNT(*) AS n FROM audit_log').get().n;
  return { rows, total, page, pageSize, totalPages: Math.ceil(total / pageSize) };
}

module.exports = { log, getForUser, getAll };